crates/legacy/src/builder_state.rs

```diff
@@ -748,9 +748,8 @@ impl<TYPES: NodeType> BuilderState<TYPES> {
         // be included, because it alone goes over sequencer's block size limit.
         // We need to drop it and mark as "included" so that if we receive
         // it again we don't even bother with it.
-        if actual_txn_count == 0 {
+        if actual_txn_count == 0 && !self.tx_queue.is_empty() {
```

Comment:
This is an interesting point. The `is_empty` check was added specifically because this logic is a little weird.  This was originally built with the assumption that if we got here at all it meant that we originally had a non-empty list of transactions, but they were pruned such that the list is now empty.

With the changes introduced in this PR, this is no longer the case. What the `is_empty` check is actually doing here, is ensuring we don't hit this early return for proposing an empty block.  The `pop_front` behavior is the original logic.  I'm not entirely certain why it's necessary, but I didn't want to change the original behavior beyond what I originally had.

Reply:
Yeah, I think it makes sense to do an empty check here, but maybe we could avoid some redundancy.

IIUC it's fine not to explicitly call `is_empty`, because `pop_front` will return `Some` only if it's not empty. We can just `return None` inside the `if let Some` clause.

I think `pop_front` is actually necessary because we want to move exactly the first element from `tx_queue` (which is too large) and insert it to `included_txns`, so that we won't try to process it anymore.

If this makes sense, then the empty check is covered by `pop_front`, which can't be removed, so we don't need `is_empty` to do the empty check again.

Reply:
That's fair enough!

Addressed in a87e6f134d38a03d3da533fa96da8c7bfe5da13f.